src/fifo.rs

```diff
@@ -0,0 +1,197 @@
+        }
+    }
+}
+
+impl FIFOConfig {
+    /// Returns values to be written to CTRL_REG9 and FIFO_CTRL:   
+    fn f_fifo_ctrl(&self) -> u8 {
+        let mut data = 0u8;
+        data |= self.fifo_mode.value();
+        data |= self.fifo_threshold;
+        data
+    }
+    fn f_ctrl_reg9(&self) -> u8 {
+        let mut data = 0u8;
+        if self.fifo_temperature_enable {
+            data |= 1 << 4;
+        }
+        if self.fifo_enable {
+            data |= 1 << 1;
+        }
+        if self.fifo_use_threshold {
+            data |= 1;
+        }
+        data
+    }
+}
+
+#[derive(Debug)]
+/// Contents of the FIFO_STATUS register (threshold reached, overrun, empty, stored data level)
+pub struct FifoStatus {
+    /// FIFO threshold status. True if FIFO filling is equal or higher than    threshold level
+    pub fifo_thresh_reached: bool,
+    /// True is FIFO is completely filled and at least one samples has been overwritten
+    pub fifo_overrun: bool,
+    /// True if FIFO is empty (level = 0)
+    pub fifo_empty: bool,
+    /// Number of unread samples stored into FIFO
+    pub fifo_level: u8,
+}
+
+impl<T> LSM9DS1<T>
+where
+    T: Interface,
+{
+    /// Enable and configure FIFO
+    pub fn configure_fifo(&mut self, config: FIFOConfig) -> Result<(), T::Error> {
+        // write values to the FIFO_CTRL register
+        self.interface.write(
+            Sensor::Accelerometer,
+            register::AG::FIFO_CTRL.addr(),
+            config.f_fifo_ctrl(),
+        )?;
+
+        // write values to specific bits of the CTRL_REG9 register
+        let ctrl_reg9: u8 =
+            self.read_register(Sensor::Accelerometer, register::AG::CTRL_REG9.addr())?;
+        let data: u8 = config.f_ctrl_reg9();
+        let mut payload: u8 = ctrl_reg9 & !FIFO_Bitmasks::CTRL_REG9_FIFO;
+        payload |= data;
+        self.interface.write(
+            Sensor::Accelerometer,
+            register::AG::CTRL_REG9.addr(),
+            payload,
+        )?;
+
+        Ok(())
+    }
+
+    /// Get flags and FIFO level from the FIFO_STATUS register
+    pub fn get_fifo_status(&mut self) -> Result<FifoStatus, T::Error> {
+        let fifo_src = self.read_register(Sensor::Accelerometer, register::AG::FIFO_SRC.addr())?;
+        let fifo_level_value = fifo_src & FIFO_Bitmasks::FSS;
+        let status = FifoStatus {
+            /// Is FIFO filling equal or higher than the threshold?
+            fifo_thresh_reached: match fifo_src & FIFO_Bitmasks::FTH {
+                0 => false,
+                _ => true,
+            },
+            /// Is FIFO full and at least one sample has been overwritten?
+            fifo_overrun: match fifo_src & FIFO_Bitmasks::OVRN {
+                0 => false,
+                _ => true,
+            },
+            /// Is FIFO empty (no unread samples)?
+            fifo_empty: match fifo_level_value {
+                0 => true,
+                _ => false,
+            },
```

Comment:
Can we use matches! macro? Maybe it is more readable?

```rust
            /// Is FIFO filling equal or higher than the threshold?
            fifo_thresh_reached: matches!(fifo_src & FIFO_Bitmasks::FTH, 1),
            /// Is FIFO full and at least one sample has been overwritten?
            fifo_overrun: matches!(fifo_src & FIFO_Bitmasks::OVRN, 1),
            /// Is FIFO empty (no unread samples)?
            fifo_empty: matches!(fifo_level_value, 0),
```

Reply:
Looks interesting, but I have never ever used it. I need to look it up :) If we decided to use it, it should be changed consistently accross the whole crate, right?

Reply:
I'm not sure about this macro, as in: I don't really understand how it works (and so far found only not-very-clear docs), and I like the more explicit code much more. I don't think it is more readable, quite the contrary. I would leave it as is.